### PR TITLE
Pydantic 2 and STAC API 1.0

### PR DESCRIPTION
**Pull request by thomas-maschler**
- Bump Pydantic to version 2.4.1
- Implement PEP 621 and use pyproject.toml 
- Implement STAC API Specs 1.0.0

This PR addresses #124 and picks up the work started in #126 which seems stale.
~~It is currently blocked by https://github.com/developmentseed/geojson-pydantic/issues/147~~

### CODE REVIEW

File: stac_pydantic/api/links.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+from stac_pydantic.utils import AutoValueEnum
+
+
+class PaginationMethods(str, AutoValueEnum):
+    """
+    https://github.com/radiantearth/stac-api-spec/blob/v1.0.0/item-search/README.md#http-request-methods-and-content-types
+    """
+
+    GET = auto()
+    POST = auto()
```
Comment (vincentsarago) [inline stac_pydantic/api/links.py:19]: We can maybe replace this enum with a Literal!

File: stac_pydantic/collection.py [hunk 2]
```diff
@@ -45,11 +45,18 @@ class Collection(Catalog):
...
-    title: Optional[str]
-    keywords: Optional[List[str]]
-    providers: Optional[List[Provider]]
-    summaries: Optional[Dict[str, Union[Range, List[Any], Dict[str, Any]]]]
-    type: str = Field("Collection", const=True, min_length=1)
+    title: Optional[str] = None
+    keywords: Optional[List[str]] = None
+    providers: Optional[List[Provider]] = None
+    summaries: Optional[Dict[str, Union[Range, List[Any], Dict[str, Any]]]] = None
+    type: str = "Collection"  # type:ignore
```
Comment (vincentsarago) [inline stac_pydantic/collection.py:55]: use Literal here

File: stac_pydantic/extensions.py [hunk 3]
```diff
@@ -1,18 +1,22 @@
...
 def validate_extensions(
-    stac_obj: Union[Item, Collection, Catalog, dict], reraise_exception: bool = False
+    stac_obj: Union[Item, Collection, Catalog, Dict[str, Any]],
+    reraise_exception: bool = False,
 ) -> bool:
     if isinstance(stac_obj, dict):
         stac_dict = stac_obj
     else:
-        stac_dict = stac_obj.dict()
+        stac_dict = json.loads(stac_obj.model_dump_json())
```
Comment (vincentsarago) [inline stac_pydantic/extensions.py:19]: why not using `stac_obj.model_dump()`?

**Comment by vincentsarago** [comment]
@thomas-maschler thanks a lot for this 🙏 , just a couple comments for now but that's a really good start.

File: stac_pydantic/catalog.py [hunk 4]
```diff
@@ -1,30 +1,30 @@
...
+    type: str = "Catalog"
+
+    model_config = ConfigDict(use_enum_values=True, extra="allow")
+
+    @field_validator("type")
+    @classmethod
+    def type_value(cls, v: str) -> str:
+        if v != "Catalog":
+            raise ValueError("Field `type` must be `Catalog`")
+        return v
```
[the remark this answers is not in the document]
Comment (thomas-maschler) [reply stac_pydantic/catalog.py]: Thanks, I override the type in collection and mypy was giving me a hard time. I got around this now, by creating a base class where `type: str` and then set different Literals in `Catalog` and `Collection`.

File: stac_pydantic/extensions.py (hunk 3 above)
Comment (thomas-maschler) [reply stac_pydantic/extensions.py:19]: Yes, I agree this looks odd. I tried this first. But `model_dump` returns Python types like `URL` and `datetime`.  `jsonschema` doesn't like this and expects the string representation.
I will add an inline comment.

**Review by iliion — APPROVED** [review]

File: tests/test_item_collection.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+PATH = ["tests", "api", "examples", f"v{STAC_API_VERSION}"]
+
+# TODO sort this out. has extra attributes but missing links
+# @pytest.mark.parametrize("example_url", [f"https://raw.githubusercontent.com/radiantearth/stac-api-spec/v{STAC_API_VERSION}/fragments/itemcollection/examples/itemcollection-sample-full.json",
+#                                          f"https://raw.githubusercontent.com/radiantearth/stac-api-spec/v{STAC_API_VERSION}/fragments/itemcollection/examples/itemcollection-sample-minimal.json"])
+# def test_item_collection_examples(example_url):
+#     """
+#     Testing the less strict version here, not enforcing required Links
+#     """
+#     compare_example(example_url, ItemCollection)
```
Comment (gadomski) [inline tests/test_item_collection.py:19]: Looks like some leftover TODOs in tests, please remove (or get to work).

File: tests/api/extensions/test_fields.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            },
+            {
+                "href": "http://collection",
+                "rel": "collection",
+            },
+        ],
+        bbox=[0, 0, 0, 0],
+        type="Feature",
+    )
+    print(fields.filter)
```
Comment (gadomski) [inline tests/api/extensions/test_fields.py:34]: Leftover print in test

**Review by gadomski — CHANGES_REQUESTED** [review]
Thanks for the contribution! In general makes sense to me -- the API structures in particular do seem to make more sense as **pydantic** models. Couple of nits about the tests to change.

At a higher level, could you add a note to the README clarifying the relationship between this library and **pystac**? My main motivation behind https://github.com/radiantearth/stac-spec/discussions/1252 is to avoid confusion in the community (and, until this PR, drop an un-maintained repo), and even a fixed up **stac-pydantic** may confuse newcomers as to "which repo should I use"? A README explanation could help that.

**Review by iliion — APPROVED** [review]

**Comment by rbavery** [comment]
curious if #127 will be merged soon? I'd like to use it in conjunction with a module I'm writing that uses Pydantic v2 to model a STAC extension for machine learning models.

It seems like there are only two nits to fix, @gadomski @thomas-maschler is it ok if I address these by removing the commented tests and the leftover print? are there other blockers to merging after that?

**Comment by thomas-maschler** [comment]
@rbavery , that would be great. I had to change focus in the meantime and I didn't get a chance to work on this again. I am not sure if I am testing the right way and removing the test probably won't do any harm.

**Comment by gadomski** [comment]
Yeah, @rbavery if those nits get cleaned up we can merge, then we'll want a release I reckon. I'm still hoping for _some_ sort of resolution in https://github.com/radiantearth/stac-spec/discussions/1252 but for now no reason not to use @thomas-maschler's work ❤️.

**Comment by thomas-maschler** [comment]
Ok, this wasn't big of an issue. There were still some STAC API specific features which were tested in the the wrong place.
@gadomski can you please glance over this again. If we were to release this, what version would this be 3.0? Should I bump the version number for `__version__`?

As for finding a resolution for https://github.com/radiantearth/stac-spec/discussions/1252 I suggest we create a separate PR and update the Readme, clarifying that the main purpose of this library is to provide reusable request/response models for tools such as fastapi. For more comprehensive schema validation and usage of extensions, pystac is the recommended tool.

I am not sure if there is consensus about the lifecycle of this library, but it would be worth mentioning, that this library will be archived eventually.

File: tests/api/test_landing_page.py [hunk 7]
```diff
@@ -57,7 +57,6 @@ def test_landing_page_invalid_features(example_url):
         LandingPage(**example)
 
 
-# TODO: Not sure if this actually does anything??
```
Comment (thomas-maschler) [inline tests/api/test_landing_page.py]: This does the right thing, and throws an error if provided a different validation schema

**Comment by gadomski** [comment]
>  If we were to release this, what version would this be 3.0? Should I bump the version number for __version__?

Probably yes, but I need to look over again to be sure. Don't bother bumping in this PR, we'll have a separate release PR.

> I suggest we create a separate PR and update the Readme, clarifying that the main purpose of this library is to provide reusable request/response models for tools such as fastapi. For more comprehensive schema validation and usage of extensions, pystac is the recommended tool.

I like this a lot 👍🏼.

**Review by gadomski — APPROVED** [review]

**Comment by gadomski** [comment]
I've created a release branch, which includes a README note about this repo's relationship to **pystac**: https://github.com/stac-utils/stac-pydantic/pull/129.

**Pull request merged**